File: tests/oracle/block_calcloop_1204.cc
/*
 * BLOCK-CALCLOOP-0001: locked Ghidra 12.0.4 BlockGraph::calcLoop
 * (block.cc:2104-2147) oracle, including the structureLoops failsafe
 * wiring (block.cc:2211-2214) and the structureReset graph chain
 * (structureLoops + calcForwardDominator, funcdata_block.cc:711-712).
 *
 * The fixture builds synthetic control-flow graphs through the production
 * BlockGraph APIs (newBlock/addEdge), then either calls calcLoop directly
 * (fresh unlabeled edges: calcLoop's own f_loop_edge writes are isolated
 * from the spanning-tree labels) or drives the full structureLoops /
 * structureLoops+calcForwardDominator chain. The post-call observation
 * covers every input the oracle consumers read: the component list order,
 * per-block index / visitcount / numdesc / mark / mark2 / copymap /
 * immed_dom, the final rootlist, and every out-edge plus mirrored in-edge
 * label.
 *
 * Cases cover: a reducible single back edge labeled f_loop_edge by the
 * direct DFS, nested inner/outer loops labeled latch-first, a self loop on
 * slot 0 followed by a fresh child on slot 1, a double calcLoop run whose
 * second pass must skip already-labeled edges (block.cc:2131 isLoopOut)
 * and leave the state identical, a diamond whose cross edge into a
 * visited-but-popped block truncates the search without a label
 * (block.cc:2138 else), the classic two-entry irreducible pair driven
 * end-to-end through structureLoops (findIrreducible labels the second
 * entry f_irreducible, irreduciblecount > 0, calcLoop then labels the
 * cycle's back edge f_loop_edge), and the structureReset chain
 * structureLoops+calcForwardDominator on a nested irreducible-inside-loop
 * graph (copymap FIND reuse) projecting the immediate dominators.
 */

#include <bits/stdc++.h>

// Test-only access: FlowBlock's private fields sit behind an explicit
// `private:` label, but BlockGraph's helper members (calcLoop,
// structureLoops, calcForwardDominator, ...) are implicitly private (no
// access label after `class BlockGraph : public FlowBlock {`), so
// `private -> public` alone cannot reach them.  `class -> struct` opens
// the implicit-private section; no decompile header uses
// `template<class ...>` or `enum class`, and <bits/stdc++.h> has already
// pulled every standard header, so the rewrite only touches Ghidra
// declarations in this translation unit.
#define private public
#define class struct
#include "architecture.hh"
#include "block.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef class
#undef private

using namespace ghidra;

static std::string flagsText(uint4 label)
{
  std::string out;
  if (label & FlowBlock::f_goto_edge) out += 'g';
  if (label & FlowBlock::f_loop_edge) out += 'l';
  if (label & FlowBlock::f_defaultswitch_edge) out += 'd';
  if (label & FlowBlock::f_irreducible) out += 'i';
  if (label & FlowBlock::f_tree_edge) out += 't';
  if (label & FlowBlock::f_forward_edge) out += 'f';
  if (label & FlowBlock::f_cross_edge) out += 'c';
  if (label & FlowBlock::f_back_edge) out += 'b';
  if (label & FlowBlock::f_loop_exit_edge) out += 'x';
  if (out.empty()) out.push_back('-');
  return out;
}

class Graph {
public:
  BlockGraph graph;
  std::vector<FlowBlock *> created;

  FlowBlock *makeBlock(void)
  {
    FlowBlock *bl = graph.newBlock();
    created.push_back(bl);
    return bl;
  }

  void edge(FlowBlock *from, FlowBlock *to)
  {
    graph.addEdge(from, to);
  }

  std::string name(FlowBlock *bl) const
  {
    std::ostringstream out;
    out << 'b' << (std::find(created.begin(), created.end(), bl) - created.begin());
    return out.str();
  }

  // Uniform post-call projection shared by every call path. `extra` carries
  // the per-case driver annotation (calcLoop pass count or "-" for the
  // structureLoops chains, whose internals never escape the driver).
  // `full_state` gates the copymap/numdesc projection: FlowBlock's
  // user-provided constructor (block.cc:61-69) initializes flags, index,
  // visitcount, parent and immed_dom but leaves copymap and numdesc
  // indeterminate until findSpanningTree initializes them (block.cc:1025-
  // 1027), so the direct calcLoop cases (no spanning tree) print "-" —
  // those values are heap noise, not algorithm output.
  void observe(const std::string &caseName, const std::string &extra, bool fullState)
  {
    std::ostringstream roots, list, blocks, outedges, inedges;
    const std::vector<FlowBlock *> &gl = graph.getList();
    for (int4 i = 0; i < gl.size(); ++i) {
      if (i != 0) list << ',';
      list << name(gl[i]);
    }
    blocks << '[';
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      if (i != 0) blocks << ';';
      blocks << name(bl) << ":index=" << bl->index
             << ",visit=" << bl->visitcount
             << ",desc=";
      if (fullState)
        blocks << bl->numdesc;
      else
        blocks << '-';
      blocks << ",mark=" << ((bl->flags & FlowBlock::f_mark) ? 1 : 0)
             << ",mark2=" << ((bl->flags & FlowBlock::f_mark2) ? 1 : 0)
             << ",copy=";
      if (!fullState)
        blocks << '-';
      else if (bl->copymap == bl)
        blocks << "self";
      else if (bl->copymap == (FlowBlock *)0)
        blocks << "null";
      else
        blocks << "other:" << name(bl->copymap);
      blocks << ",dom=";
      if (bl->immed_dom == (FlowBlock *)0)
        blocks << "null";
      else
        blocks << name(bl->immed_dom);
    }
    blocks << ']';
    bool first = true;
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      for (int4 s = 0; s < bl->outofthis.size(); ++s) {
        if (!first) outedges << ';';
        first = false;
        outedges << name(bl) << '!' << s << '>' << name(bl->outofthis[s].point)
                 << ':' << flagsText(bl->outofthis[s].label);
      }
    }
    first = true;
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      for (int4 s = 0; s < bl->intothis.size(); ++s) {
        if (!first) inedges << ';';
        first = false;
        inedges << name(bl) << '@' << s << '<' << name(bl->intothis[s].point)
                << ':' << flagsText(bl->intothis[s].label);
      }
    }
    std::cout << "case=" << caseName
              << "|extra=" << extra
              << "|list=[" << list.str() << ']'
              << "|blocks=" << blocks.str()
              << "|out=[" << outedges.str() << ']'
              << "|in=[" << inedges.str() << ']' << '\n';
  }

  // Rootlist projection for the structureLoops chains (funcdata_block.cc:
  // 711-713 consume it: calcForwardDominator input and the
  // rootlist.size() > 1 -> blocks_unreachable decision).
  void observeRoots(const std::string &caseName, const std::string &extra,
                    const std::vector<FlowBlock *> &rootlist)
  {
    std::ostringstream roots;
    for (int4 i = 0; i < rootlist.size(); ++i) {
      if (i != 0) roots << ',';
      roots << name(rootlist[i]);
    }
    std::cout << "case=" << caseName << "_roots|extra=" << extra
              << "|roots=[" << roots.str() << ']'
              << "|unreachable=" << (rootlist.size() > 1 ? 1 : 0) << '\n';
  }
};

int main(void)
{
  std::cout << std::unitbuf;
  std::vector<std::string> spec_paths;
  startDecompilerLibrary(spec_paths);
  std::cout << "schema=1|fixture=BLOCK-CALCLOOP-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    {
      // Reducible single loop b0->b1->b2->b1: direct calcLoop on the fresh
      // (unlabeled) graph. DFS path b0,b1,b2; b2!0->b1 closes the cycle
      // (b1 carries f_mark2), so addLoopEdge(b2,0) labels it f_loop_edge;
      // no other edge is touched (block.cc:2133-2137).
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b2);
      g.edge(b2, b1);
      g.graph.calcLoop();
      g.observe("reducible_backedge_labeled", "direct:1", false);
    }
    {
      // Nested loops: inner latch b3->b2 and outer latch b4->b1. The DFS
      // labels the inner back edge first, then continues to b4 and labels
      // the outer back edge; both latches keep f_loop_edge (block.cc:2121-
      // 2144 stack discipline).
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      FlowBlock *b4 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b2);
      g.edge(b2, b3);
      g.edge(b3, b2);
      g.edge(b3, b4);
      g.edge(b4, b1);
      g.graph.calcLoop();
      g.observe("nested_loops_inner_outer", "direct:1", false);
    }
    {
      // Self loop on slot 0 with a fresh child on slot 1: b1 is on its own
      // DFS path when its out-edge 0 is scanned, so the self edge is the
      // cycle (addLoopEdge(b1,0)); slot 1 still descends into b2.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b1);
      g.edge(b1, b2);
      g.graph.calcLoop();
      g.observe("self_loop_then_child", "direct:1", false);
    }
    {
      // Double calcLoop run on the reducible loop: the second pass must
      // skip the edge labeled by the first (block.cc:2131 isLoopOut
      // continue), leave the labels untouched, and clear both marks again
      // (block.cc:2145-2146) — the projection is identical to one pass.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b2);
      g.edge(b2, b1);
      g.graph.calcLoop();
      g.graph.calcLoop();
      g.observe("calc_loop_twice_loopout_skip", "direct:2", false);
    }
    {
      // Diamond b0->{b1,b2}, both into b3, with back edge b3->b1. The DFS
      // descends b0!0 first: b3!0->b1 closes a cycle (labeled). After b1
      // and b3 pop, b0!1 descends into b2 and b2!0->b3 hits a
      // visited-but-popped block (f_mark set, f_mark2 clear): the search
      // truncates with NO label and NO descent (block.cc:2138 else).
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      g.edge(b3, b1);
      g.graph.calcLoop();
      g.observe("visited_truncate_no_label", "direct:1", false);
    }
    {
      // Classic two-entry irreducible pair driven end-to-end through
      // structureLoops (block.cc:2194-2215): findSpanningTree builds the
      // tree (b0!0->b1, b1!0->b2 tree; b2!0->b1 back), findIrreducible
      // promotes the second entry b0!1->b2 to f_irreducible
      // (irreduciblecount = 1), so calcLoop runs (block.cc:2211-2214) and
      // its DFS labels the cycle's back edge b2!0->b1 f_loop_edge on top
      // of the spanning-tree labels.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b1, b2);
      g.edge(b2, b1);
      std::vector<FlowBlock *> rootlist;
      g.graph.structureLoops(rootlist);
      g.observe("irreducible_endtoend_calcloop", "loops", true);
      g.observeRoots("irreducible_endtoend_calcloop", "loops", rootlist);
    }
    {
      // structureReset graph chain (funcdata_block.cc:711-712):
      // structureLoops followed by calcForwardDominator on a nested
      // irreducible-inside-outer-loop graph (the findIrreducible FIND-reuse
      // shape). irreduciblecount > 0 so calcLoop labels both latches; the
      // dominator projection pins the chain's combined observable state
      // (immed_dom per block, entry null).
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      FlowBlock *b4 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      g.edge(b3, b2);
      g.edge(b2, b4);
      g.edge(b4, b1);
      std::vector<FlowBlock *> rootlist;
      g.graph.structureLoops(rootlist);
      g.graph.calcForwardDominator(rootlist);
      g.observe("structure_reset_chain_dominator", "reset", true);
      g.observeRoots("structure_reset_chain_dominator", "reset", rootlist);
    }
  }
  catch (const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/block_calcloop_1204.rs
// BLOCK-CALCLOOP-0001: Rugra comparand for the locked Ghidra 12.0.4
// BlockGraph::calcLoop (block.cc:2104-2147) oracle.  Mirrors
// tests/oracle/block_calcloop_1204.cc case for case: the same synthetic
// control-flow graphs are built through the production BlockGraph APIs
// (add_block/add_edge), calc_loop runs directly (fresh unlabeled edges) or
// the full structure_loops / structure_loops+calc_forward_dominator chain
// drives the failsafe path, and the complete post-call state is printed in
// the shared observation format.

use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::block::{edge_flags as ef, block_flags as bf, BlockBasic, BlockGraph, FlowBlock};

type BlockRef = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

fn flags_text(label: u32) -> String {
    let mut out = String::new();
    if label & ef::F_GOTO_EDGE != 0 { out.push('g'); }
    if label & ef::F_LOOP_EDGE != 0 { out.push('l'); }
    // Rugra's F_DEFAULTSWITCH_EDGE shares bit 7 with F_TREE_EDGE
    // (pre-existing edge_flags collision). Within this projection bit 7 can
    // only be a tree label: the spanning tree wipes every edge label at the
    // start of each pass and no pass here sets default-switch labels.
    if label & ef::F_DEFAULTSWITCH_EDGE != 0 && label & ef::F_TREE_EDGE == 0 { out.push('d'); }
    if label & ef::F_IRREDUCIBLE_EDGE != 0 { out.push('i'); }
    if label & ef::F_TREE_EDGE != 0 { out.push('t'); }
    if label & ef::F_FORWARD_EDGE != 0 { out.push('f'); }
    if label & ef::F_CROSS_EDGE != 0 { out.push('c'); }
    if label & ef::F_BACK_EDGE != 0 { out.push('b'); }
    if label & ef::F_LOOP_EXIT_EDGE != 0 { out.push('x'); }
    if out.is_empty() { out.push('-'); }
    out
}

struct Graph {
    graph: BlockGraph,
    created: Vec<BlockRef>,
    base: u64,
}

impl Graph {
    fn new(base: u64) -> Self {
        Graph { graph: BlockGraph::new(), created: Vec::new(), base }
    }

    fn make_block(&mut self) -> BlockRef {
        let bl: BlockRef = Arc::new(RwLock::new(BlockBasic::new(
            0,
            Address::new(self.base),
        )));
        self.graph.add_block(bl.clone());
        self.created.push(bl.clone());
        bl
    }

    fn edge(&mut self, from: &BlockRef, to: &BlockRef) {
        self.graph.add_edge(from.clone(), to.clone());
    }

    fn name(&self, bl: &BlockRef) -> String {
        for (i, cand) in self.created.iter().enumerate() {
            if Arc::ptr_eq(cand, bl) {
                return format!("b{}", i);
            }
        }
        "b?".to_string()
    }

    // Uniform post-call projection shared by every call path. `extra`
    // carries the per-case driver annotation, mirroring the C++ fixture.
    // `full_state` gates the copymap/numdesc projection: FlowBlock's
    // user-provided constructor (block.cc:61-69) leaves copymap and numdesc
    // indeterminate until findSpanningTree initializes them, so the direct
    // calcLoop cases print "-" (heap noise is not algorithm output).
    fn observe(&self, case_name: &str, extra: &str, full_state: bool) {
        let list: Vec<String> = self.graph.blocks.iter().map(|b| self.name(b)).collect();
        let mut blocks = String::from("[");
        for (i, bl) in self.created.iter().enumerate() {
            if i != 0 { blocks.push(';'); }
            let g = bl.read().unwrap();
            let flags = g.get_flags();
            blocks.push_str(&format!(
                "{}:index={},visit={},desc={},mark={},mark2={},copy=",
                self.name(bl),
                g.get_index(),
                g.get_visit_count(),
                if full_state { g.get_num_desc().to_string() } else { "-".to_string() },
                if flags & bf::MARK != 0 { 1 } else { 0 },
                if flags & bf::MARK2 != 0 { 1 } else { 0 },
            ));
            let (dom_weak, copy_weak) = {
                let dom = g.get_immed_dom();
                (dom, g.get_copy_map())
            };
            drop(g);
            if !full_state {
                blocks.push('-');
            } else {
                match copy_weak.and_then(|w| w.upgrade()) {
                    Some(arc) if Arc::ptr_eq(&arc, bl) => blocks.push_str("self"),
                    Some(arc) => blocks.push_str(&format!("other:{}", self.name(&arc))),
                    None => blocks.push_str("null"),
                }
            }
            match dom_weak.and_then(|w| w.upgrade()) {
                Some(arc) => blocks.push_str(&format!(",dom={}", self.name(&arc))),
                None => blocks.push_str(",dom=null"),
            }
        }
        blocks.push(']');
        let mut outedges = String::from("[");
        let mut first = true;
        for bl in &self.created {
            let g = bl.read().unwrap();
            for s in 0..g.size_out() {
                if let Some(e) = g.get_out(s) {
                    if !first { outedges.push(';'); }
                    first = false;
                    outedges.push_str(&format!(
                        "{}!{}>{}:{}",
                        self.name(bl),
                        s,
                        self.name(&e.point),
                        flags_text(e.flags)
                    ));
                }
            }
        }
        outedges.push(']');
        let mut inedges = String::from("[");
        first = true;
        for bl in &self.created {
            let g = bl.read().unwrap();
            for s in 0..g.size_in() {
                if let Some(e) = g.get_in(s) {
                    if !first { inedges.push(';'); }
                    first = false;
                    inedges.push_str(&format!(
                        "{}@{}<{}:{}",
                        self.name(bl),
                        s,
                        self.name(&e.point),
                        flags_text(e.flags)
                    ));
                }
            }
        }
        inedges.push(']');
        println!(
            "case={}|extra={}|list=[{}]|blocks={}|out={}|in={}",
            case_name,
            extra,
            list.join(","),
            blocks,
            outedges,
            inedges,
        );
    }

    // Rootlist projection for the structure_loops chains (funcdata_block.cc:
    // 711-713: calcForwardDominator input + blocks_unreachable decision).
    fn observe_roots(&self, case_name: &str, extra: &str, rootlist: &[BlockRef]) {
        let roots: Vec<String> = rootlist.iter().map(|b| self.name(b)).collect();
        println!(
            "case={}_roots|extra={}|roots=[{}]|unreachable={}",
            case_name,
            extra,
            roots.join(","),
            if rootlist.len() > 1 { 1 } else { 0 },
        );
    }
}

fn main() {
    println!("schema=1|fixture=BLOCK-CALCLOOP-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    // case=reducible_backedge_labeled
    {
        let mut g = Graph::new(0x6300);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.graph.calc_loop();
        g.observe("reducible_backedge_labeled", "direct:1", false);
    }

    // case=nested_loops_inner_outer
    {
        let mut g = Graph::new(0x6340);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        let b4 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b2, &b3);
        g.edge(&b3, &b2);
        g.edge(&b3, &b4);
        g.edge(&b4, &b1);
        g.graph.calc_loop();
        g.observe("nested_loops_inner_outer", "direct:1", false);
    }

    // case=self_loop_then_child
    {
        let mut g = Graph::new(0x6380);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b1);
        g.edge(&b1, &b2);
        g.graph.calc_loop();
        g.observe("self_loop_then_child", "direct:1", false);
    }

    // case=calc_loop_twice_loopout_skip
    {
        let mut g = Graph::new(0x63c0);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.graph.calc_loop();
        g.graph.calc_loop();
        g.observe("calc_loop_twice_loopout_skip", "direct:2", false);
    }

    // case=visited_truncate_no_label
    {
        let mut g = Graph::new(0x6400);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b3, &b1);
        g.graph.calc_loop();
        g.observe("visited_truncate_no_label", "direct:1", false);
    }

    // case=irreducible_endtoend_calcloop
    {
        let mut g = Graph::new(0x6440);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        let mut rootlist: Vec<BlockRef> = Vec::new();
        g.graph.structure_loops(&mut rootlist).expect("structure loops");
        g.observe("irreducible_endtoend_calcloop", "loops", true);
        g.observe_roots("irreducible_endtoend_calcloop", "loops", &rootlist);
    }

    // case=structure_reset_chain_dominator
    {
        let mut g = Graph::new(0x6480);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        let b4 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b3, &b2);
        g.edge(&b2, &b4);
        g.edge(&b4, &b1);
        let mut rootlist: Vec<BlockRef> = Vec::new();
        g.graph.structure_loops(&mut rootlist).expect("structure loops");
        g.graph.calc_forward_dominator(&rootlist).expect("forward dominator");
        g.observe("structure_reset_chain_dominator", "reset", true);
        g.observe_roots("structure_reset_chain_dominator", "reset", &rootlist);
    }
}
